### PR TITLE
Site Editor: use EditorProvider instead of custom logic

### PR DESCRIPTION
**Pull request by youknowriad**
Related to #52632 
Follow up to #55970

## What?

Refer to the discussion in #52632 for the reasoning. The ultimate goal is to be able to reuse more UI components (maybe the whole editor) between edit-post and edit-site.

So this PR updates the site editor to use `EditorProvider` instead of its custom `BlockEditorProvider` usage.

The biggest impact is going to be on three things which we'll have to test properly:

 - The different template/page modes in the site editor
 - Editing navigation in the site editor

### CODE REVIEW

File: packages/edit-site/src/components/block-editor/block-editor-provider/navigation-block-editor-provider.js
Comment (youknowriad) [inline packages/edit-site/src/components/block-editor/block-editor-provider/navigation-block-editor-provider.js:3]: @draganescu I'm trying to absorb the logic that we had here within the `EditorProvider` of `@wordpress/editor`. Can you help ensure I didn't break anything for the navigation pages in the site editor.

File: packages/edit-site/src/components/block-editor/block-editor-provider/default-block-editor-provider.js
Comment (youknowriad) [inline packages/edit-site/src/components/block-editor/block-editor-provider/default-block-editor-provider.js:1]: The logic that was here need to be absorbed by the `EditorProvider` from the `@wordpress/editor` package so I'm trying to add a `templateMode` prop there to support that. I added a prop but I'm not there yet, I first need to understand the logic and what are all the modes, how to trigger them...

File: packages/edit-site/src/components/block-editor/use-site-editor-settings.js
Comment (youknowriad) [inline packages/edit-site/src/components/block-editor/use-site-editor-settings.js:166]: I hope I can remove this function entirely but we can only do so when all of the site editor is relying on EditorProvider and/or the `BlockEditorProvider` usage within the site editor doesn't need the extra settings provided in the `EditorProvider`

File: packages/edit-site/src/components/editor/index.js
Comment (youknowriad) [inline packages/edit-site/src/components/editor/index.js:144]: All this logic has been moved directly to the `EditorProvider` component. It's going to benefit the post editor as well.

File: packages/edit-site/src/components/editor/index.js
Comment (youknowriad) [inline packages/edit-site/src/components/editor/index.js:194]: Here I basically just replaced these three providers with one `EditorProvider`

File: packages/editor/src/components/provider/index.js
Comment (youknowriad) [inline packages/editor/src/components/provider/index.js:45]: The assumption here is that for all editors that render a "navigation" post type, we want to put the block in "contentOnly" mode.

File: packages/editor/src/components/provider/index.js
Comment (youknowriad) [inline packages/editor/src/components/provider/index.js]: The assumption here is that for all editors that render a "navigation" post type, we want to create a wrapper block to ensure everything is rendered properly. The wrapper is "locked" but we can edit its content in contentOnly mode.

File: packages/editor/src/components/provider/index.js
Comment (youknowriad) [inline packages/editor/src/components/provider/index.js]: The assumption here is that queryContext is just local state, am I wrong here? cc @ntsekouras

File: packages/editor/src/components/provider/index.js
Comment (ramonjd) [inline packages/editor/src/components/provider/index.js:185]: Just noticed while testing editing pages in the editor that the post object isn't resolved in time so the properties `id` and `type` might need some optional chaining or something.

**Review by ramonjd — COMMENTED** [review]
There's something I couldn't yet work out when testing. When opening a page from the Design menu like this (don't edit anything, just navigate to a page and try to edit):

https://github.com/WordPress/gutenberg/assets/6458278/e70ea157-ff29-45db-bfcc-2754dde4a155

The `editor.BlockEdit` filters are not firing, namely the ones that determine the content edit mode in [withDisableNonPageContentBlocks](https://github.com/WordPress/gutenberg/blob/57211ac3d2d05e67378cb471dd77140cf680b1e0/packages/edit-site/src/components/page-content-focus-manager/disable-non-page-content-blocks.js#L42-L42).

The consequence is that I can't edit the post-content.

I didn't get to the bottom of it, but I noticed that when first editing a page this way, it looks like the `blocks` array returned from `useEntityBlockEditor()` is empty in [edit-template](https://github.com/WordPress/gutenberg/pull/56000/files#diff-c9814423e8df6d31ebbf8c155b117a537d85f113245688549d9275cc4c6208ccR65), so maybe something to do with the edited record in that hook. Sorry I ran out of time!

Tricky thing is that if I open another template first, then navigate to a page to edit, then it's fine so I can only guess some context is loading at this stage. 🤔

File: packages/editor/src/components/provider/index.js
Comment (ntsekouras) [reply packages/editor/src/components/provider/index.js]: I think this is obsolete. I can open a PR to remove it, as it requires a couple more changes in block's edit function.

Comment (ntsekouras) [reply packages/editor/src/components/provider/index.js]: PR that removes it here: https://github.com/WordPress/gutenberg/pull/56034

**Review by draganescu — COMMENTED** [review]
i tested the change only around the navigation section of the site editor (this PR removes the navigation provider, which is great) - so far things work as expected with one exception: 

- make sure to have one navigation block which only contains a page list
- navigating to site editor -> navigation
- selecting the navigation with page list
- clicking edit
- the page list block when selected shows two edit buttons for some reason

<details>
<summary>Screenshot</summary>
Screenshot 2023-11-13 at 05 13 31

<details>

**Comment by youknowriad** [comment]
@draganescu I haven't been able to replicate the duplicated "edit" button.

**Comment by youknowriad** [comment]
@ramonjd I think there's a pre-existing issue with these filters. I wonder if it's a bug in `withFilters` component. Basically I see the filter being added but the components are not re-rendered with the said filter. So maybe something within withFilters.

But Also, the way we're adding and removing these filters dynamically doesn't feel great to me to be honest.

**Comment by youknowriad** [comment]
The last failing test is one that you've added @draganescu (focus loss on template part) I haven't been able to understand why it fails in this PR.

**Review by andrewserong — COMMENTED** [review]
Just catching up, and I see that https://github.com/WordPress/gutenberg/pull/56000/commits/572dd729ea258882d0fce9263e85239cc3c2233f removes the logic for the hidden template mode introduced in https://github.com/WordPress/gutenberg/pull/52674 that grabs the real blocks used within the template. This means that the hidden template mode no longer shows the post featured image block if present (and doesn't factor in the order in which the content blocks appear in the template):

| Trunk | This PR |
| --- | --- |
| image | image |

Would it be worth preserving that behaviour so that the content matches the order used in the template? I think the main benefit of the prior behaviour is probably the visibility of the featured image block when present in the template.

---

Other than that, this is testing well for me so far!

File: packages/edit-site/src/components/block-editor/use-site-editor-settings.js
Comment (andrewserong) [inline packages/edit-site/src/components/block-editor/use-site-editor-settings.js:101]: Tiny nit: these are both still returned by the `useSelect` even though they're not retrieved in this destructuring. Is it worth removing them from within the `useSelect` callback, too (lines 133 and 134), since the corresponding lines are now handled in `useSiteEditorSettings`?

File: packages/edit-site/src/components/editor/index.js
Comment (andrewserong) [inline packages/edit-site/src/components/editor/index.js:231]: Moving this `Notice` changes where in the UI it's displayed. It now sits above the interface (including both sidebars), rather than within the editor canvas. Is that intentional? It looks decent to me, and I don't have a strong opinion, but thought I'd mention it since I wasn't sure if this PR was meaning to make visual changes 🙂

I'm not too sure how to trigger this warning to display manually, so locally I just made the condition `true` in order to render it:

| Before | This PR |
| --- | --- |
| image | image |

**Comment by youknowriad** [comment]
> Would it be worth preserving that behaviour so that the content matches the order used in the template? I think the main benefit of the prior behaviour is probably the visibility of the featured image block when present in the template.

I can restore it (with a bit of code complexity) but to be honest, I'm not sure we should do, or either why this mode should be different than the post editor? In the post editor, the featured image is visible in the sidebar instead?

File: packages/edit-site/src/components/editor/index.js
Comment (youknowriad) [reply packages/edit-site/src/components/editor/index.js:231]: That was not my intention. How did you reproduce this? For me that message shows up only if the canvas doesn't show up (! editedPost) condition. So maybe I didn't understand it properly.

**Comment by youknowriad** [comment]
One thing I noticed and that could be improved but probably in follow-ups is that the "locked" mode is actually handled within components in edit-site while it could (and probably should) be handled within the EditorProvider

File: packages/edit-site/src/components/editor/index.js
Comment (andrewserong) [reply packages/edit-site/src/components/editor/index.js:231]: I couldn't actually get it to fire manually through real means, so my screenshot is somewhat artificial (via removing the condition). If it's looking fine for you when `! editedPost` is true, then that sounds good to me!

**Comment by andrewserong** [comment]
> I can restore it (with a bit of code complexity) but to be honest, I'm not sure we should do, or either why this mode should be different than the post editor? In the post editor, the featured image is visible in the sidebar instead?

That's a good question, this is a tricky one to figure out! It might be worth getting some design input on it, but I think the feature was built as part of addressing some of the ideas from the great unification issue (https://github.com/WordPress/gutenberg/issues/41717) (the PR https://github.com/WordPress/gutenberg/pull/52674 mentions item 3 from that issue). I suppose the question would be: what's the desired way for folks to interact with the featured image block in a `post-only` mode? My understanding was that by exposing the featured image block within the list view and editor canvas, it made it (potentially) easier for folks to find and interact with.

For the purposes of this PR, perhaps another way to ask the question might be: from the perspective of `useBlockEditorProps`, should the post-only mode template be dynamically generated (based on page content blocks found within a template) or inferred and use a hard-coded "template" as in the current state of this PR? Perhaps a middle ground might be to construct that hard-coded template based on which `PAGE_CONTENT_BLOCK_TYPES` are found, similar to how the site editor calls `__experimentalGetGlobalBlocksByName` before determining whether to show the template preview toggle.

I think I'd lean toward either looking at scaling back the template preview toggle feature in a separate PR (to unblock this PR), or trying to get this PR to match the behaviour on `trunk`, so that this PR doesn't try to do too many changes at once.

All up, I don't feel strongly about which approach to go with. I'm not sure how helpful all that is, but I'm happy to keep testing, whichever way you'd like to go with this!

**Comment by youknowriad** [comment]
I've restored the template preview mode like trunk (extract the blocks from the template) in [c34e322](https://github.com/WordPress/gutenberg/pull/56000/commits/c34e32287f5ce96980217eabfbf2a7312f50aad8). You can see the complexity in the commit and why I'm not sure it's worth it but I'm definitely ok with it for now.

**Comment by youknowriad** [comment]
I bootstrapped some documentation about the `EditorProvider` component. There's a few things that stood out to me while writing the documentation, that would be good API changes but I didn't want to make breaking changes at the moment or deprecations (maybe later):

 - `post` prop: Why are we expecting a full object here, a couple post type, post id should be enough.
 - `__unstableTemplate` prop: We should probably deprecate this prop entirely because we can easily guess the template from `core-data`. I don't think we need to support rendering any random template here (we don't do that anywhere anyway) 
 - `mode` prop: I like this new prop, it clarifies a lot of things but right now, it's not implemented fully within the `EditorProvider`. Some locking/unlocking behavior is still within edit-site and should be moved at some point. 


With that I think the PR is ready to be considered for merge.

File: packages/editor/src/components/provider/README.md
Comment (ramonjd) [inline packages/editor/src/components/provider/README.md]:
```suggestion
This is the rendering mode of the post editor. We support multiple rendering modes:
```

File: packages/editor/src/components/provider/README.md
Comment (ramonjd) [inline packages/editor/src/components/provider/README.md]:
```suggestion
-   `all`: This is the default mode. It renders the post editor with all the features available. If a template is provided, it's preferred over the post.
```

File: packages/edit-site/src/components/sidebar-edit-mode/page-panels/edit-template.js
Comment (ramonjd) [inline packages/edit-site/src/components/sidebar-edit-mode/page-panels/edit-template.js]:
```suggestion
				_postType,
```

File: packages/editor/src/components/provider/index.js
Comment (ramonjd) [inline packages/editor/src/components/provider/index.js]: Should this be `{string}` ?

File: packages/editor/src/components/provider/index.js
Comment (ramonjd) [inline packages/editor/src/components/provider/index.js]:
```suggestion
		if ( blocks[ i ].innerBlocks.length ) {
			result.push(
				...extractedPageContentBlockTypesFromTemplateBlocks(
					blocks[ i ].innerBlocks
				)
			);
		}
```

Might save 0.000001 of a nanosecond 😄 and prevent pushing empty arrays to results

File: packages/editor/src/components/provider/index.js
Comment (ramonjd) [inline packages/editor/src/components/provider/index.js]: Optional: remove premature multiblock checking and replace with string comparison, e.g.,  `'core/query' === blocks[ i ].name`

**Review by ramonjd — APPROVED** [review]
Great work! 

The site editor is working as expected for me:

✅ I can edit navigation post types
✅ Editing a page I can perform expected actions: toggling template editing mode, swapping templates. 
✅ Template previews appear as expected. I can edit a page's template and the changes persist between preview modes (template hidden and displayed)
✅  Command center commands look good: testing editing template, editing various templates/pages

Also can't find any regressions when viewing or editing styles, patterns or templates.

Would be good to get a second review though. 

> You can see the complexity in the commit and why I'm not sure it's worth it but I'm definitely ok with it for now.

To maintain consistency between template and page editing in the site editor, from a user-facing perspective, I'd say it's worth maintaining it. There might be a case for a refactor, or even a review of the feature in general. It's pretty buried.

File: packages/editor/src/components/provider/index.js
Comment (andrewserong) [inline packages/editor/src/components/provider/index.js]: Tiny nit (feel free to ignore since this function instead exported): would the imperative `extract` be clearer than `extracted`?

**Review by andrewserong — APPROVED** [review]
Wonderful, great follow-ups @youknowriad, thanks for preserving the behaviour on `trunk` for now and for adding in the extra documentation 👍

Nothing to add from me beyond what @ramonjd already mentioned. All testing nicely:

✅ Navigation editing is working correctly 
✅ All site, template, page, post content with and without template preview modes are working correctly
✅ Code changes look good

> You can see the complexity in the commit and why I'm not sure it's worth it but I'm definitely ok with it for now.

Yes, I see what you mean. Fortunately it looks like it'll be fairly straightforward to remove in follow-ups if we decide to scale back / simplify or remove the template preview toggle. As Ramon mentioned, it could be good to review the feature in general and figure out the desired level of prominence 👍

LGTM! ✨

**Comment by youknowriad** [comment]
Thanks folks for the reviews, I appreciate it.

**Comment by youknowriad** [comment]
Looks like this PR made the site editor navigation 3 or 4 times better and slowed a little bit the initial loading. I call that a win.

**Comment by youknowriad** [comment]
The dev note should clarify that now the "editor" store is actually filled with a real entity and not just an empty useless store.

**Comment by andrewserong** [comment]
I think this has introduced a regression for the list view when editing a Navigation entity in the site editor, or when rearranging blocks within the `post-only` mode, due to `createBlock` being called too frequently in the `useMemo`. I have a potential fix up in #56666

**Pull request merged**
